update exif format replacing queues.

tsteven4 opened the pull request:
So far I have chosen not to pursue a shift from pointers to references to minimize the changes in the diff.

tsteven4 committed:
replace exif format usage of queues.

tsteven4 committed:
fix exif test to run in any time zone.

tsteven4 committed:
fix new exif memory leaks.

robertlipe commented on exif.cc:
```diff
@@ -440,24 +414,25 @@ exif_read_ifd(exif_app_t* app, const uint16_t ifd_nr, gbsize_t offs,
     tag->count = gbfgetuint32(fin);
     tag->size = exif_type_size(tag->type) * tag->count;
-    tag->data = &tag->value;
+    tag->data.reset(reinterpret_cast<char *>(xcalloc(4, 1)), xfree);
```
ick

robertlipe commented on exif.cc:
```diff
@@ -440,24 +414,25 @@ exif_read_ifd(exif_app_t* app, const uint16_t ifd_nr, gbsize_t offs,
+      gbfread(tag->data.data(), 4, 1, fin);
     } else {
       tag->value = gbfgetuint32(fin);
+      memcpy(tag->data.data(), &tag->value, 4);
```
I'd have to think about this really hard for big-endian machines, but I'm not sure we care about them much any more.

tsteven4 replied:
This is a bug, and it manifests itself not based on the endianness of the machine, but based on the endianness of the image as stored in the TIFF header.  The big endian byte order isn't too common, but I found a sample image we can use to test.

tsteven4 replied:
There is also a different preexisting bug here!

robertlipe commented on exif.cc:
```diff
@@ -570,15 +544,15 @@ exif_read_app(exif_app_t* app)
+exif_examine_app(ExifApp* app)
 {
-  gbfile* ftmp = exif_app->fcache;
+  gbfile* ftmp = app->fcache;
```
So the original was looking at a global exif_app instead of the function argument? Wow.

robertlipe commented on exif.cc:
```diff
@@ -620,14 +592,13 @@ exif_find_ifd(exif_app_t* app, const uint16_t ifd_nr)
-    queue* elem, *tmp;
-    QUEUE_FOR_EACH(&ifd->tags, elem, tmp) {
-      exif_tag_t* tag = reinterpret_cast<exif_tag_t *>(elem);
+    for (auto& tag_instance : ifd->tags) {
```
This could probably be a smarter type to avoid a linear scan, but you've already poked this pretty hard in this CL so that can wait another day.

tsteven4 replied:
sorry, bigger fish to fry.

robertlipe commented on exif.cc:
```diff
@@ -844,15 +814,15 @@ exif_waypt_from_exif_app(exif_app_t* app)
-    } else if (memcmp(tag->data, "UNICODE\0", 8) == 0) {
+    if (memcmp(tag->data.data(), "ASCII\0\0\0", 8) == 0) {
+      wpt->notes = QString::fromLatin1(tag->data.data() + 8, tag->size - 8);
+    } else if (memcmp(tag->data.data(), "UNICODE\0", 8) == 0) {
```
Both of these tests could probably be just plain ole operator ==. I'm not sure I care about testing trailing junk in a data structure that's both null terminated and fixed length. "ASCII\0\0377\07" is probably a legit match, no?

(I also totally get the mindset of faithfully converting existing code instead of rethinking it.)

tsteven4 replied:
These exact strings, including all the nulls, are specified in Table 9 and 35 of the EXIF specification.  Also, the user comment itself may not be null terminated, although the tag->data may be depending on how we implement it.  
(http://www.cipa.jp/std/documents/e/DC-008-Translation-2016-E.pdf)

tsteven4 replied:
There is a bug just below this, unicode can be UTF-16BE or UTF-16LE depending on file endianness.

robertlipe commented on exif.cc:
```diff
@@ -969,41 +939,38 @@ exif_put_value(const int ifd_nr, const uint16_t tag_id, const uint16_t type, con
     }
 
-    tag = (exif_tag_t*) xcalloc(sizeof(*tag), 1);
+    ifd->tags.append(ExifTag());
```
Would it be more natural to have a local that's populated and appended instead of one that's created and immediately stuffed into a queue and then pulled back out?

tsteven4 replied:
It would be more natural to me as well, and I actually started down that path.  But in the end I went with less disruptive more literal translation.  Sometimes a pointer to a list element is returned and we need to get it back from the list anyway.

robertlipe commented on exif.cc:
```diff
@@ -969,41 +939,38 @@ exif_put_value(const int ifd_nr, const uint16_t tag_id, const uint16_t type, con
-        tag->data = tmp;
-        tag->data_is_dynamic = 1;
-      }
+      auto* tmp = reinterpret_cast<char *>(xmalloc(tag->size < 4 ? 4 : tag->size));
```
This makes my head hurt. This whole mess makes me think that tag->data should be a std::vector<uint8_t> or something that's just resized as we go or something. Maybe that surgery is too radical for now...But QSharedPointers wrapping xmalloc/xfree seems like the wrong long-term solution.

tsteven4 replied:
QByteArray drops in for QSharedPointer<char> almost transparently.  There is still a ton of char manipulation through .data() or .constData().  QByteArray does clean up the messy replacement for xrealloc cited here though.

robertlipe commented on exif.cc:
```diff
@@ -1251,9 +1208,9 @@ exif_write_apps()
-      QUEUE_FOR_EACH(&app->ifds, e1, t1) {
-        exif_ifd_t* ifd = reinterpret_cast<exif_ifd_t *>(e1);
-        sortqueue(&ifd->tags, exif_sort_tags_cb);
+      for (auto& ifd_instance : app->ifds) {
```
Could this loop be combined with the one above or does GPS_IFD interfere with the sort? Thinking about readability here(if that's not a lost cause in this module :-) not the performance const of repeating it.)

tsteven4 replied:
maybe, but it will have to wait.

robertlipe commented on exif.cc:
```diff
@@ -1263,16 +1220,10 @@ exif_write_apps()
-        exif_ifd_t* ifd = reinterpret_cast<exif_ifd_t *>(e1);
-        exif_ifd_t* ifd_next = reinterpret_cast<exif_ifd_t *>(t1);
-        char next;
+      for (int i=0; i<app->ifds.size(); ++i) {
```
Whitespace

tsteven4 replied:
done.

robertlipe reviewed (commented):
I wonder if .data really should be a std::vector<uint8_t> or a QByteArray or something - anything - other than a char* that's reallocated and copied endlessly. Maybe it's an array of QVariants so you can get bools and doubles and chars in and out without endianness and alignment hassles.

But to the point of this CL, eradicating queue and using QList, I think that made this much more readable, so thanx!

tsteven4 committed:
incorporate review suggestions and fix bugs.

correct user comment encoding based on image endianness.
fix possible tag corruption when data fits in 4 bytes depending
on image endianness and processor endianness.

tsteven4 committed:
fix codacy/cppcheck whining.

GPSBabelDeveloper reviewed (approved):
Even with the grubbiness we've agreed to walk away from, this is *so much* more type-safe and generally readable, that it's a poster child fo such cleanup. Thanx for going the extra mile and improving test coverage and having the tenacity to dig up the specs and find that it really is a fixed-length AND zero terminated buffer, for example.

Thanx for a nice batch of fixes!

tsteven4 committed:
improvements to exif format.

fix bugs with rational types including greatest common divisor.
turn off assertions for release builds in GPSBabel.pro.
clean up exif.

tsteven4 committed:
use continued fractions to create TIFF RATIONAL types.

some references were regenerated as the new algorithm
can generate different values for the RATIONAL numerator
and denominator, in this case 0.0 -> 0/INT32_MAX.

tsteven4 committed:
use QVariant to store exif tag data.

enhance exif debug capability.

tsteven4 committed:
minor fixes for exif.

tsteven4 committed:
replace legacy time handling in exif.

also add support of exif OffsetTime* tags.

tsteven4 committed:
fix exif warnings.

tsteven4 committed:
fix exif dynamic debug omissions.

tsteven4 committed:
add exif support for uncompressed thumbnails.
